src/Stratis.SmartContracts.Executor.Reflection/StateTransitionResult.cs

```diff
@@ -34,9 +70,104 @@ public class StateTransitionResult
         /// </summary>
         public uint160 ContractAddress { get; }
 
+    }
+
+    /// <summary>
+    /// Represents the result of a failed state transition.
+    /// </summary>
+    public class StateTransitionError
+    {
+        public StateTransitionError(Gas gasConsumed, StateTransitionErrorKind kind, Exception vmException)
+        {
+            this.Kind = kind;
+            this.GasConsumed = gasConsumed;
+            this.VmException = vmException;
+        }
+
+        /// <summary>
+        /// An exception thrown by the VM. This value is null unless <see cref="Kind"/>
+        /// equals <see cref="StateTransitionErrorKind.VmError"/> or <see cref="StateTransitionErrorKind.OutOfGas"/>.
+        /// </summary>
+        public Exception VmException { get; }
+
         /// <summary>
-        /// Whether the operation was successful.
+        /// The kind of error that occurred during the state transition.
         /// </summary>
-        public bool Success { get; }
+        public StateTransitionErrorKind Kind { get; }
+
+        /// <summary>
+        /// The gas consumed during execution.
+        /// </summary>
+        public Gas GasConsumed { get; }
+    }
+
+    /// <summary>
+    /// The result of a state transition operation.
+    /// </summary>
+    public class StateTransitionResult
+    {
+        private StateTransitionResult(StateTransitionError error)
+        {
+            this.IsSuccess = false;
+            this.Error = error;
+        }
+
+        private StateTransitionResult(StateTransitionSuccess success)
+        {
+            this.IsSuccess = true;
+            this.Success = success;
+        }
+
+        /// <summary>
+        /// The gas consumed during the state transition.
+        /// </summary>
+        public Gas GasConsumed => this.IsSuccess ? this.Success.GasConsumed : this.Error.GasConsumed;
+
+        public bool IsSuccess { get; }
+
+        public bool IsFailure => !this.IsSuccess;
+
+        /// <summary>
+        /// The successful result of the state transition. This value is null unless
+        /// <see cref="IsSuccess"/> equals true.
+        /// </summary>
+        public StateTransitionSuccess Success { get; }
```

Comment:
This is interesting. Not opposed to it, but I think we should rename this then, if we have `IsSuccess` as well.